### PR TITLE
Push secret to worker

### PR DESCRIPTION
**Pull request by protatremy**
Related to the issue PushSecretToWorker step #3197
This pullrequest is implementing the two steps needed by the push secret functionality to populate secrets in files in a worker directory:
- create files in the worker dir
- delete files

## Contributor Checklist:

* [x] I have updated the unit tests
* [ ] I have created a file in the master/buildbot/newsfragment directory (and read the README.txt in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/steps/push_secret_to_worker.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from __future__ import print_function
+
+import os
+
+from buildbot.process.buildstep import SUCCESS
+from buildbot.process.buildstep import BuildStep
+from buildbot.process.buildstep import LoggingBuildStep
+
+
+class workerFileSecret():
```
Comment (tardyp) [inline master/buildbot/steps/push_secret_to_worker.py]: class should start with Capital letter, and derive from object

File: master/buildbot/steps/push_secret_to_worker.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    def __init__(self, populated_secret_list, **kwargs):
+        super(PushSecretToWorker, self).__init__(**kwargs)
+        self.secret_to_be_populated = []
+        for secretvalue, path in populated_secret_list:
+            self.secret_to_be_populated.append(workerFileSecret(secretvalue, path))
+
+    def runPopulateSecrets(self):
+        for secret in self.secret_to_be_populated:
+            secret.createSecretFile()
```
Comment (tardyp) [inline master/buildbot/steps/push_secret_to_worker.py]: this method does not exist.
We need to separate secret value from secret population mechanism.

PushSecretToWorker needs to inherit from buildsteps.worker.CompositeStepMixin
you need to add their a method to upload a file to worker

File: master/buildbot/steps/push_secret_to_worker.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    def __init__(self, paths, **kwargs):
+        self.paths = paths
+        super(RemoveWorkerFileSecret, self).__init__(**kwargs)
+
+    def runRemoveWorkerFileSecret(self):
+        for path in self.paths:
+            if not os.path.exists(path):
+                raise ValueError("Path %s does not exist")
+            os.remove(path)
```
Comment (tardyp) [inline master/buildbot/steps/push_secret_to_worker.py]: that needs to happen on the worker, here this is on the master!

File: master/buildbot/test/unit/test_push_secret_to_worker.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        tempdir = FilePath(self.mktemp())
+        tempdir.createDirectory()
+        return tempdir.path
+
+    def setUp(self):
+        self.temp_path = self.createTempDir("tempdir")
+
+    def testPushSecretToWorkerStepSuccess(self):
+        self.setupStep(PushSecretToWorker([(os.path.join(self.temp_path, "pathA"), "something")]))
+        self.expected_remote_commands = ""
```
Comment (tardyp) [inline master/buildbot/test/unit/test_push_secret_to_worker.py]: you should actually test that remote commands are created which create the files

File: master/buildbot/steps/push_secret_to_worker.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+
+class workerFileSecret():
+
+    def __init__(self, path, secretvalue):
+        self.path = path
+        if not isinstance(self.path, str):
+            raise ValueError("Secret path %s is not a string" % path)
+        self.secretvalue = secretvalue
+        if not os.path.exists(os.path.dirname(self.path)):
```
Comment (tardyp) [inline master/buildbot/steps/push_secret_to_worker.py]: you cannot make this check as this is on the worker

File: master/buildbot/steps/download_secret_to_worker.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(self, path, secretvalue):
+        self.path = path
+        if not isinstance(self.path, str):
+            raise ValueError("Secret path %s is not a string" % path)
+        self.secretvalue = secretvalue
+        if not os.path.exists(os.path.dirname(self.path)):
+            raise ValueError("Path %s does not exist")
+
+
+class DownloadSecretsToWorker(_TransferBuildStep, CompositeStepMixin):
```
Comment (tardyp) [inline master/buildbot/steps/download_secret_to_worker.py]: I dont think it needs to inherit transferbuildstep. BuildStep should be enough + CompositeStepMixin of course

File: master/buildbot/steps/download_secret_to_worker.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            res = yield self.downloadFileContentToWorker(secret.path, secret.secretvalue)
+            all_results.append(res)
+        if FAILURE in all_results:
+            result = FAILURE
+        else:
+            result = SUCCESS
+        defer.returnValue(result)
+
+    @defer.inlineCallbacks
+    def start(self):
```
Comment (tardyp) [inline master/buildbot/steps/download_secret_to_worker.py]: this is old style step. please don't use, and rather use run()

File: master/buildbot/steps/download_secret_to_worker.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        for path in self.paths:
+            res = yield self.runRmFile(path, abandonOnFailure=False)
+            all_results.append(res)
+        if FAILURE in all_results:
+            result = FAILURE
+        else:
+            result = SUCCESS
+        defer.returnValue(result)
+
+    @defer.inlineCallbacks
```
Comment (tardyp) [inline master/buildbot/steps/download_secret_to_worker.py]: ditto (run)

File: master/buildbot/test/unit/test_download_secret_to_worker.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        tempdir.createDirectory()
+        return tempdir.path
+
+    def setUp(self):
+        self.temp_path = self.createTempDir("tempdir")
+
+    def testPushSecretToWorkerStepSuccess(self):
+        self.setupStep(DownloadSecretsToWorker([(os.path.join(self.temp_path, "pathA"),
+         "something")]))
+        self.expected_remote_commands = ""
```
Comment (tardyp) [inline master/buildbot/test/unit/test_download_secret_to_worker.py]: I would expect a command here!

File: master/buildbot/test/unit/test_download_secret_to_worker.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        self.exp_result = SUCCESS
+        self.runStep()
+
+    def testPathDoesNotExists(self):
+        self.assertRaises(ValueError,
+                          lambda: self.setupStep(DownloadSecretsToWorker([(os.path.join("/dir/pathA"),
+                                                                           "something")])))
+
+
+class TestDownloadFileSecretToWorker(steps.BuildStepMixin, unittest.TestCase):
```
Comment (tardyp) [inline master/buildbot/test/unit/test_download_secret_to_worker.py]: it is unclear to me what is test difference there is with previous testcase that would deserve another test/name

File: master/buildbot/test/util/steps.py [hunk 11]
```diff
@@ -417,13 +417,11 @@ def _remotecommand_run(self, command, step, conn, builder_name):
         self.assertEqual(step, self.step)
         self.assertEqual(conn, self.conn)
         got = (command.remote_command, command.args)
-
         if not self.expected_remote_commands:
```
Comment (tardyp) [inline master/buildbot/test/util/steps.py]: I think those line deletions are unneeded

File: worker/buildbot_worker/commands/fs.py [hunk 12]
```diff
@@ -261,3 +261,23 @@ def start(self):
             self.sendStatus(
                 {'header': '%s: %s: %s' % (self.header, e.strerror, dirname)})
             self.sendStatus({'rc': e.errno})
+
+
+class RemoveFile(base.Command):
```
Comment (tardyp) [inline worker/buildbot_worker/commands/fs.py]: arg.. its annoying to require a change in the slave.
This means that people will need to upgrade.

We at least need to:
- increment the worker version
- if the worker version is < 3.1, then we should use another method. probably overwrite the file with empty string, or use rmdir command (which looks like baked by rm -rf on unix)
.

File: master/buildbot/steps/download_secret_to_worker.py (hunk 6 above)
Comment (protatremy) [inline master/buildbot/steps/download_secret_to_worker.py]: ok

File: master/buildbot/steps/download_secret_to_worker.py (hunk 7 above)
Comment (protatremy) [inline master/buildbot/steps/download_secret_to_worker.py]: ok

File: master/buildbot/steps/download_secret_to_worker.py (hunk 8 above)
Comment (protatremy) [inline master/buildbot/steps/download_secret_to_worker.py]: ok

File: master/buildbot/test/unit/test_download_secret_to_worker.py (hunk 9 above)
Comment (protatremy) [inline master/buildbot/test/unit/test_download_secret_to_worker.py]: ok

File: master/buildbot/test/unit/test_download_secret_to_worker.py (hunk 10 above)
Comment (protatremy) [inline master/buildbot/test/unit/test_download_secret_to_worker.py]: simplified

File: master/buildbot/test/util/steps.py (hunk 11 above)
Comment (protatremy) [inline master/buildbot/test/util/steps.py]: done

File: worker/buildbot_worker/commands/fs.py (hunk 12 above)
Comment (protatremy) [inline worker/buildbot_worker/commands/fs.py]: done

**Pull request merged**